### ISSUE TITLE
Zone: NgZone should allow an application to install custom hooks into onTurnStart/Done.

### ISSUE THREAD

**Issue by matanlurey** [angular/angular#1978]
In the past, it was possible to hook into `VmTurnZone` (now renamed `NgZone`) by wrapping onTurnStart/Done with a custom callback that called the Angular registered function, then your own. This allowed things like syncing with non-Angular services or for profiling.

Now the only method is `initCallbacks`, which doesn't give access to the existing callbacks. It's possible to get around this by reading into Angular's source code and doing things like calling `lifeCycle.tick()` yourself if you use `initCallbacks`, but ideally this would be provided by Angular.

**Comment by rkirov** [comment]
As commented with Matan, the best API for this would do a bit more than add a callback to onTurnDone.

The user callback should be called only after an async event has happened, change detection has ran *and* it did not schedule any more microtasks. Something like 

```
class Lifecycle {
  Promise<null> whenStable()
}
```

// cc @vicb thoughts (as the last person who worked on Zones)?

**Comment by vicb** [comment]
`whenStable()` seems like a good proposal. It probably needs modification to zones (both dart & js) as there is currently no way to know if `onTurnDone` result in enqueueing more microtask - it should be an easy update.

**Comment by juliemr** [comment]
See also https://github.com/angular/angular/issues/428

**Comment by vicb** [comment]
What we probably wants to do in [`NgZone`](https://github.com/angular/angular/blob/598a75ec1c837376b95adb3924a116531eec9d08/modules/angular2/src/core/zone/ng_zone.ts#L174)

```
                if (ngZone._pendingMicrotasks == 0 && ngZone._nestedRun == 0 &&
                    !this._inVmTurnDone) {
                  if (ngZone._onTurnDone && ngZone._hasExecutedCodeInInnerZone) {
                    try {
                      this._inVmTurnDone = true;
                      parentRun.call(ngZone._innerZone, ngZone._onTurnDone);
                    } finally {
                      this._inVmTurnDone = false;
                      ngZone._hasExecutedCodeInInnerZone = false;
                    }
                  }
                }
```

->

```
                if (ngZone._pendingMicrotasks == 0 && ngZone._nestedRun == 0 &&
                    !this._inVmTurnDone) {
                  if (ngZone._onTurnDone && ngZone._hasExecutedCodeInInnerZone) {
                    try {
                      this._inVmTurnDone = true;
                      parentRun.call(ngZone._innerZone, ngZone._onTurnDone);
// +
                      if (ngZone._pendingMicrotasks == 0) {
                        parentRun.call(ngZone._innerZone, ngZone._onStable);  
                      }
// +
                    } finally {
                      this._inVmTurnDone = false;
                      ngZone._hasExecutedCodeInInnerZone = false;
                    }
                  }
                }
```

- do the same in Dart,
- add tests

**Comment by rkirov** [comment]
It would be nice to detect timers being added through the createTimer and createPeriodicTimer APIs https://api.dartlang.org/apidocs/channels/stable/dartdoc-viewer/dart:async.ZoneDelegate 
and not call the onStable until the non-period timers have resolved.

**Comment by rkirov** [comment]
Also I think the client API should be exposed through the Lifecycle object, not NgZone.

**Comment by vicb** [comment]
> It would be nice to detect timers being added through the createTimer and createPeriodicTimer APIs https://api.dartlang.org/apidocs/channels/stable/dartdoc-viewer/dart:async.ZoneDelegate 
and not call the onStable until the non-period timers have resolved.

I guess we would only take timers that fire in a "near" future into account ?

> Also I think the client API should be exposed through the Lifecycle object, not NgZone.

Agree

**Comment by juliemr** [comment]
Closing in favor of #2808
